slothy/core/core.py

```diff
@@ -2756,6 +2756,16 @@ def _add_constraints_loop_optimization(self):
                 if self.config.sw_pipelining.max_pre < relpos < 1:
                     # pylint:disable=singleton-comparison
                     self._Add(t.pre_var == False)
+                    
+        # Forbid emerging dependencies across the loop boundary if the
+        # loop boundary (that SLOTHY pretends does not exist) uses temporary registers
+        for (consumer,producer, _, _, _, _, alloc) in self._iter_dependencies_with_lifetime():
+            for r in self.config.sw_pipelining.boundary_reserved_regs:
+                if r not in alloc.keys():
+                    continue
+                if producer.src.is_virtual is True or consumer.is_virtual is True:
+                    continue
+                self._Add(alloc[r] == False).OnlyEnforceIf(producer.src.pre_var, consumer.pre_var.Not())
```

Comment:
This needs a bit more thought and tests to be robust in general:
* There will also be cross-boundary dependencies from non-post to post
* There can be dependencies across loop iterations (say an accumulator being changed in every iteration) that need to be handled correctly as well